### PR TITLE
Accurate Xilinx 7 series architecture

### PR DESCRIPTION
**Pull request by WhiteNinjaZ**
#### Description

This PR provides the most up to date arch file for the Xilinx 7 series chip set. This PR is a WIP and is expected to change often.

### CODE REVIEW

**Comment by vaughnbetz** [comment]
@WhiteNinjaZ @jgoeders  : I think this is the most recent 7-series capture.  If it is, I think it would be good to merge into the master so we have an architecture (which can still be refined) on the master branch that is the current best capture. It would also be good to add a CI test or two targeting it to make sure it doesn't get broken.

**Comment by WhiteNinjaZ** [comment]
Yes this is the most up to date arch!

**Comment by vaughnbetz** [comment]
@WhiteNinjaZ, @jgoeders  : @MohamedElgammal and I are going through the VTR 9 paper and trying to make sure Mohamed's 7-series figure is accurate, and that the capture is fairly accurate.

1. Are all 'stubs' vertical?  I see only stub_y in the architecture (length 1 V wires).
2. It looks like CLBs have their pins on all 4 sides.  Is this the best capture, or would it be better to have pins on only two sides (say top and right)?  The INT blocks might be better modeled with a 2-sided architecture, as each CLB would connect to a single vertical and single horizontal channel, which is closer to a separate INT block I think.  However, I think Xilinx has direct connect to all 4 sides (actually can reach 8 CLBs with no wiring), which might be better modeled with pins on all 4 sides.  Are the diagonal length 1 + 1 wires capturing some of these direct connects?

**Comment by vaughnbetz** [comment]
Thanks for the fast and detailed answers @WhiteNinjaZ .  That all makes sense.  
@MohamedElgammal : I guess that means we should either not show the extra stubs in the figure, or show them vertically.

**Comment by vaughnbetz** [comment]
@WhiteNinjaZ : can this PR be merged?  It has a couple of CI failures -- are they spurious?  

I think we need to get this merged ASAP.

**Comment by WhiteNinjaZ** [comment]
@vaughnbetz: I am putting the finishing touches on the last version of this architecture and should be done in the next few days. I believe the CI failures are spurious since this PR does not make any changes to the VTR code base and only adds a new architecture.

**Comment by vaughnbetz** [comment]
OK, maybe you just need to rebase (golden results might have changed).
Even if it isn't perfect, I think we should merge this very soon (this
week).

Vaughn

**Comment by WhiteNinjaZ** [comment]
@vaughnbetz just pushed the most updated architecture: note that we are still running a few tests to try and figure out the weird increase in wirelength for the designs and if needed will make a few final changes to this architecture.

File: vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml [hunk 1]
```diff
@@ -0,0 +1,3396 @@
...
+      - Various custom TCL scripts in Vivado for proper mapping of routing structures to the VTR arch
+      file.
+
+    Routing structures approximated in this architecture:
+      - The various distributions of different wire lengths in the horizontal and vertical directions
+      are accurate to the Xilinx 7-series. Additionally, the different channel widths in
+      vertical/horizontal directions are accurate to the Xilinx 7-series (124 for horizontal 
+      and 190 for vertical).
+      - The types of wires that each wire length can connect to are accurate to the Xilinx 7-series
+      (i.e. L1 wires may only drive L1 and L2 wires) Accept for a few of the diagonal wires and
```
Comment (vaughnbetz) [inline vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:24]: Accept -> except

File: vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml [hunk 2]
```diff
@@ -0,0 +1,3396 @@
...
+      - The types of wires that each wire length can connect to are accurate to the Xilinx 7-series
+      (i.e. L1 wires may only drive L1 and L2 wires) Accept for a few of the diagonal wires and 
+      feedback connections (as noted in the custom switchblock section). However, the equations 
+      for how these wires connect to each other are not accurate and are loosely based off of the 
+      wilton switch blocks.
+      - This architecture demonstrates the use of L shaped wire types by connecting two separate wires
+      together using a short in the custom switchblocks. It is worth noting that the given channel width 
+      must be fixed at 190 or a multiple of 190 in order to avoid some of the diagonal wires 
+      becoming unconnected.
+      - Stubs (a wires ability to connect to a SB other than the one it terminates in) are modeled in
```
Comment (vaughnbetz) [inline vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:32]: wires -> wire's

File: vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml [hunk 3]
```diff
@@ -0,0 +1,3396 @@
...
+      this architecture. For Stubs who connect to SBs parallel to the same direction of travel we use the
```
Comment (vaughnbetz) [inline vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:33]: who -> which

File: vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml [hunk 4]
```diff
@@ -0,0 +1,3396 @@
...
+      built in SB pattern feature. For Stubs who connect to SBs perpendicular to the direction of travel
```
Comment (vaughnbetz) [inline vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:34]: who -> which

Comment (vaughnbetz) [reply vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:34]: (not sure why Stubs is sometimes capitalized)

File: vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml [hunk 5]
```diff
@@ -0,0 +1,3396 @@
...
+      - In Xilinx, and as mentioned in the VTR 9 paper, the concept of how many SB and CB a wire has
+      access to is essentially equal to an equivalent VTR wire of +1 length (i.e. an L1 wire in Xilinx 
+      is equivalent to an L2 in VTR). This is due to VTR having SBs offset from logic elements, while 
+      Xilinx has SBs directly connected to logic elements.
+      - The fc values for CLB inputs and outputs are close approximations to the Xilinx 7-series. 
+      Due to the large number of different wire types in the 7-series we found the VTR struggled to 
+      route designs with low minW when using a fractional fc value (i.e. a fc of 15% connecting to a 
+      wire type that only takes up 6% of the channels has no connections at W < 100). To solve this 
+      we used an absolute fc value which although less accurate, ensures that all wire types have
+      at least some connections at low channel widths. The Fc values given for other logic elements
```
Comment (vaughnbetz) [inline vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:46]: other logic elements -> other types of blocks

File: vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml [hunk 6]
```diff
@@ -0,0 +1,3396 @@
...
+      - Accurate internal structure (LUT, FF, etc.). One exception is that the actual 7-series has a tap
+      point for each COUT of each adder along the carry chain. These where not included due to VTR 
+      causing errors any time they where implemented. Also our CLB capture does not include MUX7 and MUX8
+      modes of the LUTs.
+      - Accurate partial crossbar and feedback connections
+      - Includes special bounce pips which allows CLB inputs and outputs to connect to longer wire types 
+      than the normal L1 and L2.
+    BRAM: 
+      - Acurate timing
+      - Most commonly used modes are implemented with the exception of alternate port width modes on
```
Comment (vaughnbetz) [inline vtr_flow/arch/xilinx/7series_BRAM_DSP_cary.xml:64]: alternate port width modes on single port rams -> different widths on the two ports of a simple dual port RAM.

(I don't think you mean single port here)

**Review by vaughnbetz — COMMENTED** [review]
Some suggested minor fixes to the comment at the top.
Also, the filename has a typo I think: cary should be carry

**Comment by vaughnbetz** [comment]
If you make these changes and remove the WIP we can merge this.
Is there a test for the architecture?  If not, please make one or file an issue to make one.  It can just test the 6 VTR circuits at the fixed channel width, with very loose QoR bounds.

**Comment by vaughnbetz** [comment]
Thanks @WhiteNinjaZ (and @jgoeders).  We should merge this very soon .. the VTR paper is pretty close to done. We can talk about it today, but the 7-series arch is now one of a very small number of outstanding items, so we should merge this and close out that section ASAP.

**Comment by jgoeders** [comment]
@vaughnbetz Sounds good. I'm out of town today and can't make the meeting, but it sounds like Joshua is done with the paper and just working on getting this PR through. I'll give our section another read through tomorrow to double check everything.

**Comment by vaughnbetz** [comment]
Looks good. I also suggest adding a single small circuit  (in another test suite) to vtr_reg_strong so it runs on the github runners, and is tested more frequently (even if the google CI is down).  I'd make that one a fixed channel width run, again to keep the runtime low (and please post what the runtime is for that design).  Ideally this would be a design that runs in seconds or tens of seconds.

**Comment by vaughnbetz** [comment]
One failure, but it is a known problem with odd-even routing unit tests (@soheilshahrouz : please disable that one).  Merging this and filing an issue for the strong (small circuit) test.

**Pull request merged**